Find on page intermittently fails to show results in PDFs in Safari

rr-codes opened the pull request:
#### dd929cd085dc7862b052721523e7b6ba97b98769
<pre>
Find on page intermittently fails to show results in PDFs in Safari
<a href="https://bugs.webkit.org/show_bug.cgi?id=251311">https://bugs.webkit.org/show_bug.cgi?id=251311</a>
rdar://100155151

Reviewed by Aditya Keerthi.

When viewing a PDF in Safari, Find in Page fails after the following sequence
of events:

1. Open a PDF in Safari
2. Activate Find in Page and search for a string
3. Refresh the page
4. Activate Find in Page again

After the last step, Find In Page will continue to return zero results regardless
of the search string.

This is because the `WKPDFView` instance is used as the `UIFindSession`&apos;s &quot;searchable object&quot;
to process the search. When the page is refreshed, a new `WKPDFView` instance is created, but
not a new `UIFindSession`, and so the session has lost its searchable object.

This PR fixes this by setting the session&apos;s searchable object to the new `WKPDFView` in the
`_didCommitLoadForMainFrame` method, which is called after the new `WKPDFView` has been created.
This ensures that the session&apos;s searchable object is up to date.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _didCommitLoadForMainFrame]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPage.mm:
(swizzledPerformTextSearchWithQueryString):
(TEST):

Canonical link: <a href="https://commits.webkit.org/259655@main">https://commits.webkit.org/259655@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/df5ec1250ce442a7b7f5afe22adcafd3870e422c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105486 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14591 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38399 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114746 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174897 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109385 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15936 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5819 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97792 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114598 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111243 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12170 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95157 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39660 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94034 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26793 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81352 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7871 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28149 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8224 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4742 "Found 1 new test failure: fast/images/avif-as-image.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13999 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47692 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6668 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9898 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -4091,6 +4091,9 @@ - (void)findInteraction:(UIFindInteraction *)interaction didEndFindSession:(UIFi
         [_contentView didEndTextSearchOperation];
+
+    if ([session isKindOfClass:[UITextSearchingFindSession class]])
+        [(UITextSearchingFindSession *)session setSearchableObject:[self _searchableObject]];
```
Is there a way we can fix this without using SPI? It also seems strange that this logic runs when the find session "ends".

I am wondering if there is something we can call to trigger `findInteraction:sessionForView:` to recreate the session entirely.

rr-codes replied:
> Is there a way we can fix this without using SPI?

I don't believe so; I looked through the API for various alternatives but couldn't find any. 

> It also seems strange that this logic runs when the find session "ends".

Ya, I know. It's because this is also called when the session changes. ("Called when the current search session has changed or ended")

> I am wondering if there is something we can call to trigger findInteraction:sessionForView: to recreate the session entirely.

Not that I know of, since the docs say it is "Called when a find session is requested to begin by the user"

pxlcoder replied:
Richard and I confirmed that this approach is necessary due to a limitation of the existing UIKit API.

pxlcoder reviewed (commented):
Can we add an API test for this?

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -818,8 +818,19 @@ - (void)_didCommitLoadForMainFrame
+        if (!textSearchingFindSession)
+            return;
+
+        [textSearchingFindSession setSearchableObject:[self _searchableObject]];
```
I would write this as:

```
if (auto *findSession = dynamic_objc_cast<UITextSearchingFindSession>([_findInteraction activeFindSession]))
    findSession.searchableObject = [self _searchableObject];
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPage.mm:
```diff
@@ -913,4 +913,44 @@ static void swizzledPerformTextSearchWithQueryString(id, SEL, NSString *, UIText
+        [webView loadRequest:request];
+        [webView _test_waitForDidFinishNavigation];
+
+        auto* findInteraction = [webView findInteraction];
```
`auto *` since this is an Objective-C pointer.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPage.mm:
```diff
@@ -913,4 +913,44 @@ static void swizzledPerformTextSearchWithQueryString(id, SEL, NSString *, UIText
+        [findInteraction presentFindNavigatorShowingReplace:NO];
+        [webView waitForNextPresentationUpdate];
+
+        auto* findSession = [findInteraction activeFindSession];
```
`auto *` since this is an Objective-C pointer.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPage.mm:
```diff
@@ -913,4 +913,44 @@ static void swizzledPerformTextSearchWithQueryString(id, SEL, NSString *, UIText
+    hasPerformedTextSearchWithQueryString = true;
+}
+
+TEST(WebKit, WKPDFViewFindInteractionAfterPageRefresh)
```
`FindInPDFAfterReload`? This would match then naming scheme of the other find-in-PDF test.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPage.mm:
```diff
@@ -913,4 +913,44 @@ static void swizzledPerformTextSearchWithQueryString(id, SEL, NSString *, UIText
+{
+    // Swizzle out the method that performs searching, since PDFHostViewController (a remote view
+    // (controller) cannot be created in TestWebKitAPI, and we cannot actually search the PDF.
+    std::unique_ptr<InstanceMethodSwizzler> isInBackgroundSwizzler = makeUnique<InstanceMethodSwizzler>(NSClassFromString(@"WKPDFView"), @selector(performTextSearchWithQueryString:usingOptions:resultAggregator:), reinterpret_cast<IMP>(swizzledPerformTextSearchInPDFWithQueryString2));
```
`isInBackgroundSwizzler` -> `performTextSearchInPDFWithQueryStringSwizzler`.

I must have incorrectly copied the name elsewhere, so you can update the other location too.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPage.mm:
```diff
@@ -913,4 +913,44 @@ static void swizzledPerformTextSearchWithQueryString(id, SEL, NSString *, UIText
 
+static bool hasPerformedTextSearchWithQueryString = false;
+
+static void swizzledPerformTextSearchInPDFWithQueryString2(id, SEL, NSString *, UITextSearchOptions *, id<UITextSearchAggregator> aggregator)
```
Can we just update the implementation of `swizzledPerformTextSearchInPDFWithQueryString` rather than creating a `2` that's almost the exact same?

pxlcoder reviewed (commented):
The commit message needs to be updated to reflect the new approach + files changed.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPage.mm:
```diff
@@ -883,6 +886,40 @@ static void swizzledPerformTextSearchWithQueryString(id, SEL, NSString *, UIText
     testPerformTextSearchWithQueryStringInWebView(webView.get(), @"Birthday", searchOptions.get(), 0UL);
```
We should add `hasPerformedTextSearchWithQueryString = false;` here to leave things in the right state.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPage.mm:
```diff
@@ -883,6 +886,40 @@ static void swizzledPerformTextSearchWithQueryString(id, SEL, NSString *, UIText
 
+TEST(WebKit, FindInPDFAfterReload)
+{
+    hasPerformedTextSearchWithQueryString = false;
```
You can remove this line once making the adjustment above.

rr-codes committed:
Find on page intermittently fails to show results in PDFs in Safari
https://bugs.webkit.org/show_bug.cgi?id=251311
rdar://100155151

Reviewed by Aditya Keerthi.

When viewing a PDF in Safari, Find in Page fails after the following sequence
of events:

1. Open a PDF in Safari
2. Activate Find in Page and search for a string
3. Refresh the page
4. Activate Find in Page again

After the last step, Find In Page will continue to return zero results regardless
of the search string.

This is because the `WKPDFView` instance is used as the `UIFindSession`'s "searchable object"
to process the search. When the page is refreshed, a new `WKPDFView` instance is created, but
not a new `UIFindSession`, and so the session has lost its searchable object.

This PR fixes this by setting the session's searchable object to the new `WKPDFView` in the
`_didCommitLoadForMainFrame` method, which is called after the new `WKPDFView` has been created.
This ensures that the session's searchable object is up to date.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _didCommitLoadForMainFrame]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPage.mm:
(swizzledPerformTextSearchWithQueryString):
(TEST):

Canonical link: https://commits.webkit.org/259655@main
